h2o-persist-gcs/src/main/java/water/persist/PersistGcs.java

```diff
@@ -0,0 +1,193 @@
+package water.persist;
+
+import com.google.cloud.ReadChannel;
+import com.google.cloud.storage.*;
+import water.H2O;
+import water.Key;
+import water.MemoryManager;
+import water.Value;
+import water.api.FSIOException;
+import water.fvec.FileVec;
+import water.fvec.GcsFileVec;
+import water.util.Log;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.nio.ByteBuffer;
+import java.util.*;
+
+/**
+ * Persistence backend for GCS
+ */
+public final class PersistGcs extends Persist {
+
+  private static final Storage storage = StorageOptions.getDefaultInstance().getService();
+
+  @Override
+  public byte[] load(final Value v) throws IOException {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    
+    final byte[] contentBytes = MemoryManager.malloc1(v._max);
+    final ByteBuffer wrappingBuffer = ByteBuffer.wrap(contentBytes);
+    final Key k = v._key;
+    long offset = 0;
+    // Skip offset based on chunk number
+    if(k._kb[0] == Key.CHK) {
+      offset = FileVec.chunkOffset(k); // The offset
+    }
+    
+    final ReadChannel reader = storage.reader(blobId);
+    reader.seek(offset);
+    reader.read(wrappingBuffer);
+    
+    return contentBytes;
+  }
+
+  @Override
+  public Key uriToKey(URI uri) throws IOException {
+    final GcsBlob blob = GcsBlob.of(uri);
+    final Long contentSize = storage.get(blob.getBlobId()).getSize();
+    return GcsFileVec.make(blob.getCanonical(), contentSize);
+  }
+
+  @Override
+  public void store(Value v) throws IOException {
+    if (!v._key.home()) return;
+    final byte payload[] = v.memOrLoad();
+    final GcsBlob blob = GcsBlob.of(v._key);
+    Log.debug("Storing: " + blob.toString());
+    final ByteBuffer buffer = ByteBuffer.wrap(payload);
+    storage.create(blob.getBlobInfo()).writer().write(buffer);
+  }
+
+  @Override
+  public void delete(Value v) {
+    final BlobId blobId = GcsBlob.of(v._key).getBlobId();
+    Log.debug("Deleting: " + blobId.toString());
+    storage.get(blobId).delete();
+  }
+
+  @Override
+  public void cleanUp() {
+    throw H2O.unimpl();
+  }
+
+  private static final Map<String, StringCache> keyCache = new HashMap<>();
+  private static final StringCache bucketCache = new StringCache() {
```

Comment:
Same issues with thread-safety?

Also: PersistGCS is effectively a singleton. There is no need to use static variables.

Reply:
Thread safety issued removed. Static modifier removed as well ! Thank you :)